### DISCUSSION TITLE
Local-Search behaviour / profile interpretation

### DISCUSSION THREAD

**Discussion by sschnug** [google/or-tools#3914] [Routing (and legacy CP) questions]
Looking at the following output, i'm wondering _how to interpret that_.

```shell
Local search operator statistics:
      | Neighbors | Filtered | Accepted | Time (s)
Cross |       618 |      618 |        0 | 1.3e+02
Total |       618 |      618 |        0 | 1.3e+02
Local search filter statistics (PATH_CHEAPEST_ARC):
                             |     Calls |   Rejects | Time (s) | Rejects/s
   DimensionFilter(Capacity) |   1252161 |   1246000 |   1e+05  |      12
        VariableDomainFilter |      6165 |         0 | 6.7e+02  |       0
       VehicleVariableFilter |      6164 |         0 | 6.7e+02  |       0
DimensionFilter(WorkingTime) |      6163 |         0 | 6.7e+02  |       0
 DimensionFilter(TravelTime) |      6162 |         0 | 6.7e+02  |       0
 PathCumulFilter(TravelTime) |      6161 |         0 | 6.7e+02  |       0
                       Total |   1282976 |   1246000 | 1.1e+05  |      12
Local search filter statistics:
                             |     Calls |   Rejects | Time (s) | Rejects/s
 DimensionFilter(TravelTime) |       618 |         0 | 1.1e+02  |       0
 PathCumulFilter(TravelTime) |       618 |         0 | 1.1e+02  |       0
       VehicleVariableFilter |       618 |         0 | 1.1e+02  |       0
   DimensionFilter(Capacity) |       618 |         0 | 1.1e+02  |       0
          SumObjectiveFilter |       618 |         0 | 1.1e+02  |       0
        VariableDomainFilter |       618 |         0 | 1.1e+02  |       0
DimensionFilter(WorkingTime) |       618 |         0 | 1.1e+02  |       0
                       Total |      4326 |         0 | 7.8e+02  |       0
```

It seems, during LS-phase (after first sol), there is **no reason for rejection**, at least according to those shown.

Neither Filters on `TravelTime` nor `SumObjectiveFilter` *do reject*. Same goes for `VariableDomainFilter`.

Yet, every single neighbor is rejected (here: cross-only). What could be the reason for rejection in this case?

-----

This behaviour showed up when i put (structured) **holes in travel-time domains** (`~50% dead`), think horizon = multiple days and only parts of those days are "open" in time:

```cpp
/*
valid-ranges
{[72000,108000)[151200,194400)[237600,280800)[324000,367200)[410400,450000)}
forbidden-ranges aka domain-holes = horizon -= valid
{[0,72000)[108000,151200)[194400,237600)[280800,324000)[367200,410400)[450000,486000)}
*/
VLOG(3) << "punch holes into domain";
const int64_t index = manager->NodeToIndex(SOME_FUNCTION());

for(const auto& range : domain_holes) {
  VLOG(3) << absl::StrFormat("hole: [%d, %d]", boost::icl::first(range), boost::icl::last(range));
  auto& dim = model->GetDimensionOrDie("TravelTime");
  dim.CumulVar(index)->RemoveInterval(boost::icl::first(range), boost::icl::last(range));
}
```

I can imagine, that there will be some CP-propagation/shifting necessary and i see that in (perf-based) profiling. Also that it's costly.

But it's way too costly. The difference in *new improved sols per 10 minutes* is ~ 10 vs. 10,000 or something like that. It's unclear if the solver thinks there are none or just don't find the ones. 

**In short: domain-holes in travel-time kill local-search progress and i'm unsure why exactly.**
